README.md
# ocaml-iri
OCaml implementation of Internationalized Resource Identifiers (IRIs)

RFC 3987: http://tools.ietf.org/html/rfc3987

This implementation does not depend on regular expression library.
Is is implemented using Sedlex (https://github.com/alainfrisch/sedlex),
thus it will be usable in Javascript (with Js_of_ocaml).

`OCaml-RDF` will use this library instead of its `Rdf_iri` module. 


Changes.txt
Next release:
- fix Iri.append_path: remove last empty string of path, if present,
  before appending (to avoid having // in path)

0.4.0:
- move to Uutf 1.0.0 and Uunf 2.0.0
- improve error messages when encountering invalid utf-8 characters
- install iri.cmti

0.3.0:
- upgrade to OCaml 4.03.0

0.2.0:
- fixes
- better error messages
- additions and changes in API

test_parser.ml
(*********************************************************************************)
(*                OCaml-IRI                                                      *)
(*                                                                               *)
(*    Copyright (C) 2016 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

open Iri

let prerr_endline =
  if Unix.isatty Unix.stderr then
    fun str -> prerr_endline
      (Printf.sprintf "\027[1;31m%s\027[0m" str)
  else
    prerr_endline

let test_string str =
  try
    let iri = Iri.of_string str in
    let str2 = Iri.to_string iri in
    print_endline
      (Printf.sprintf "%s -> %s %s%s" str str2
       (
        match Iri.query iri with
        | None -> ""
        | Some _ ->
            let map = Iri.query_kv iri in
            String.concat ", "
              ((List.map
                (fun (k, v) -> Printf.sprintf "%S->%S" k v))
               (Iri.KV.bindings map))
       )
      (try
         let iri = Iri.of_string str2 in
         let str3 = Iri.to_string iri in
         if str2 = str3 then
           ""
         else
           (Printf.sprintf " [not reparsed the same way str3=%s]" str3)
       with
       | e ->
           let msg =
             match e with
             | Iri_lexer.Error e ->
                 Printf.sprintf "%s"  (Iri_lexer.string_of_error e)
             | Iri.Error e ->  Iri.string_of_error e
             | e -> Printexc.to_string e
           in
           Printf.sprintf " [reparse failed: %s]" msg
      ))
  with
  | Iri_lexer.Error e ->
      prerr_endline (Printf.sprintf "%s: %s" str
       (Iri_lexer.string_of_error e))
  | Iri.Error e ->  prerr_endline (Iri.string_of_error e)
  | e -> prerr_endline (Printexc.to_string e)

let test_mine () =
  let iris =
    [ "" ;
      "urn:example:mammal:monotreme:echidna";
      "http://ab.cde123fgh.ij/kl/mn/op.html" ;
      "http:/kl/mn/op.html" ;
      "http:kl/mn/op.html" ;
      "http:?coucou=trois#id" ;
      "http://a/b/c/d/?coucou=trois#id" ;
      "http://a/b/c/d?coucou=trois#id" ;
      "ssh://foo@bar.net.com:8080/my/path/?arg=1%3D4%32&foo=15#id";
      "http://ab.גדהוזח.ij/kl/mn/op.html";
      "http://www.ほんとうにながいわけのわからないどめいんめいのらべるまだながくしないとたりない.w3.mag.keio.ac.jp/";
      "http://ほんとうにながいわけのわからないどめいんめいのらべるまだながくしないとたりない.ほんとうにながいわけのわからないどめいんめいのらべるまだながくしないとたりない.ほんとうにながいわけのわからないどめいんめいのらべるまだながくしないとたりない.w3.mag.keio.ac.jp/";
    ]
  in
  List.iter test_string iris

let test_ref_resolve () =
  let base = Iri.of_string "http://a/b/c/d;p?q" in
  let test (str, expected) =
    try
      let expected = Iri.of_string expected in
      let r = Iri.of_string str in
      let resolved = Iri.resolve ~normalize: true ~base r in
      let ok = Iri.to_string resolved = Iri.to_string expected in
      let msg =
        Printf.sprintf "%s + %s => %s [%s]"
          (Iri.to_string base)
          (Iri.to_string r)
          (Iri.to_string resolved)
          (if ok then "OK" else Printf.sprintf "KO, expected %s" (Iri.to_string expected))
      in
      if ok then print_endline msg else prerr_endline msg
    with
      Iri.Error e ->
        let msg = Printf.sprintf "%s, %s: %s" expected str (Iri.string_of_error e) in
        prerr_endline msg
  in
  List.iter test
    [
      "g:h"       ,  "g:h" ;
      "g"         ,  "http://a/b/c/g" ;
      "./g"       ,  "http://a/b/c/g" ;
      "g/"        ,  "http://a/b/c/g/" ;
      "/g"        ,  "http://a/g" ;
      "//g"       ,  "http://g" ;
      "?y"        ,  "http://a/b/c/d;p?y" ;
      "g?y"       ,  "http://a/b/c/g?y" ;
      "#s"        ,  "http://a/b/c/d;p?q#s" ;
      "g#s"       ,  "http://a/b/c/g#s" ;
      "g?y#s"     ,  "http://a/b/c/g?y#s" ;
      ";x"        ,  "http://a/b/c/;x" ;
      "g;x"       ,  "http://a/b/c/g;x" ;
      "g;x?y#s"   ,  "http://a/b/c/g;x?y#s" ;
      ""          ,  "http://a/b/c/d;p?q" ;
      "."         ,  "http://a/b/c/" ;
      "./"        ,  "http://a/b/c/" ;
      ".."        ,  "http://a/b/" ;
      "../"       ,  "http://a/b/" ;
      "../g"      ,  "http://a/b/g" ;
      "../.."     ,  "http://a/" ;
      "../../"    ,  "http://a/" ;
      "../../g"   ,  "http://a/g" ;
      "../../../g",  "http://a/g" ;
      "../../../../g",  "http://a/g" ;
       "/./g"    , "http://a/g" ;
      "/../g"    , "http://a/g" ;
      "g."       , "http://a/b/c/g." ;
      ".g"       , "http://a/b/c/.g" ;
      "g.."      , "http://a/b/c/g.." ;
      "..g"      , "http://a/b/c/..g" ;
       "./../g"    ,  "http://a/b/g" ;
      "./g/."     ,  "http://a/b/c/g/" ;
      "g/./h"     ,  "http://a/b/c/g/h" ;
      "g/../h"    ,  "http://a/b/c/h" ;
      "g;x=1/./y" ,  "http://a/b/c/g;x=1/y" ;
      "g;x=1/../y",  "http://a/b/c/y" ;
      "g?y/./x"   ,  "http://a/b/c/g?y/./x" ;
      "g?y/../x"  ,  "http://a/b/c/g?y/../x" ;
      "g#s/./x"   ,  "http://a/b/c/g#s/./x" ;
      "g#s/../x"  ,  "http://a/b/c/g#s/../x" ;
      "http:g"    ,  "http:g" ;
    ]

let test_http_link () =
  let f str =
    let l = Iri.parse_http_link str in
    let b = Buffer.create 256 in
    Printf.bprintf b "Link: %s\n" str ;
    List.iter
      (fun (s, iri) ->
         Printf.bprintf b "=> %s rel=%S\n" (Iri.to_string iri) s)
      l;
    print_endline (Buffer.contents b)
  in
  List.iter f
    [ {|<http://www.w3.org/ns/ldp#BasicContainer>; rel="type", <http://www.w3.org/ns/ldp#Resource>; rel="type" |} ;
      {|  <http://www.w3.org/ns/ldp#IndirectContainer>; rel="type",  <http://www.w3.org/ns/ldp#Resource>; rel="type"|} ;
    ]


let test_file file =
  let ic = open_in file in
  (try while true do test_string (input_line ic) done
   with End_of_file -> ()
  );
  close_in ic

let () =
  if Array.length Sys.argv > 1 then
    Array.iteri (fun i f -> if i > 0 then test_file f) Sys.argv
  else
    (test_ref_resolve () ; test_mine () ; test_http_link ())




iri.ml
(*********************************************************************************)
(*                OCaml-IRI                                                      *)
(*                                                                               *)
(*    Copyright (C) 2016 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)


type error = Iri_types.error
exception Error = Iri_types.Error

open Iri_types

let string_of_error = function
| Parse_error (str,e) ->
    let msg =
      match e with
        Iri_lexer.Error e -> Iri_lexer.string_of_error e
      | _ -> Printexc.to_string e
    in
    Printf.sprintf "Parse error in %s\n%s" str msg

let parse_error str e = raise (Error (Parse_error (str, e)))

type iri = t
module Ord = struct type t = iri let compare = Iri_types.compare end
module Set = Set.Make(Ord)
module Map = Map.Make(Ord)

let of_lexbuf ?pctdecode ?pos ?normalize lexbuf =
  let iri = Iri_lexer.iri ?pctdecode ?pos lexbuf in
  match normalize, is_relative iri with
    Some true, _
  | None, false -> Iri_types.normalize iri
  | _ -> iri

let of_string ?pctdecode ?pos ?normalize str =
  let lexbuf = Sedlexing.Utf8.from_string str in
  try of_lexbuf ?pctdecode ?pos ?normalize lexbuf
  with (Iri_lexer.Error _ ) as e -> parse_error str e

let resolve ?(normalize=true) ~base iri =
  let resolved =
    match is_relative iri with
      false -> iri
    | true ->
        let str = to_string iri in
        (*prerr_endline
          (Printf.sprintf "%s=\nscheme=%s\nhost=%s\npath=%s"
           str (scheme iri)
             (match host iri with None -> "" | Some s -> s)
             (path_string iri)
          );*)
        let len = String.length str in
        if len <= 0 then
          base
        else
          match String.get str 0 with
          | '#' -> with_fragment base (fragment iri)
          | '?' ->
              let base = with_fragment base None in
              with_query base (query iri)
          | _ ->
              let base = with_query base None in
              let base = with_fragment base None in
              let base =
                match host iri with
                  None -> base
                | Some host ->
                    let base = with_host base (Some host) in
                    let base = with_port base (port iri) in
                    with_user base (user iri)
              in
              let path =
                match path iri with
                  Absolute p -> Absolute p
                | Relative iri_path ->
                    let path =
                      match path base with
                        Absolute [] -> Absolute iri_path
                      | Relative [] -> Relative iri_path
                      | Absolute l ->
                          let l = List.rev iri_path @ (List.tl (List.rev l)) in
                          Absolute (List.rev l)
                      | Relative l ->
                          let l = List.rev iri_path @ (List.tl (List.rev l)) in
                          Absolute (List.rev l)
                    in
                    Iri_types.path_remove_dot_segments path
              in
              let i = with_path base path in
              let i = with_fragment i (fragment iri) in
              let i = with_query i (query iri) in
              i
  in
  if normalize then Iri_types.normalize resolved else resolved

let parse_http_link str =
  let lexbuf =
    try Sedlexing.Utf8.from_string str
    with Sedlexing.MalFormed as e ->
      raise (Error (Parse_error ("Malformed character in http link: "^str, e)))
  in
  Iri_lexer.http_link lexbuf

let to_uri = Iri_types.to_uri
let normalize = Iri_types.normalize
let with_fragment = Iri_types.with_fragment
let fragment = Iri_types.fragment
let query_set = Iri_types.query_set
let query_opt = Iri_types.query_opt
let query_get = Iri_types.query_get
let with_query_kv = Iri_types.with_query_kv
let with_query = Iri_types.with_query
let query_kv = Iri_types.query_kv
let query = Iri_types.query
let append_path = Iri_types.append_path
let path_string = Iri_types.path_string
let with_path = Iri_types.with_path
let path = Iri_types.path
let with_port = Iri_types.with_port
let port = Iri_types.port
let with_host = Iri_types.with_host
let host = Iri_types.host
let with_user = Iri_types.with_user
let user = Iri_types.user
let with_scheme = Iri_types.with_scheme
let scheme = Iri_types.scheme
let to_string = Iri_types.to_string
let equal = Iri_types.equal
let compare = Iri_types.compare
let is_relative = Iri_types.is_relative
let is_absolute = Iri_types.is_absolute
let iri = Iri_types.iri
type t = Iri_types.t
type path = Iri_types.path = Absolute of string list | Relative of string list
type query_kv = Iri_types.query_kv
module KV = Iri_types.KV


iri_types.ml
(*********************************************************************************)
(*                OCaml-IRI                                                      *)
(*                                                                               *)
(*    Copyright (C) 2016 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

type path =
| Absolute of string list
| Relative of string list

module KV = Map.Make(String)
type query_kv = string KV.t

type t = {
    scheme : string ;
    user : string option ;
    host : string option ;
    port : int option ;
    path : path ;
    query : string option ; (** not %-decoded as query is not parse to name/value pairs *)
    fragment : string option ;
    mutable query_kv : query_kv  option ;
      (** key value pairs from query string to avoid parsing it various times *)
  }

type error =
| Parse_error of string * exn

exception Error of error

let is_absolute t =
  match t.fragment with
    None -> t.scheme <> ""
  | _ -> false

let is_relative t = t.scheme = ""

let utf8_nb_bytes_of_char c =
  let n = Char.code c in
  if n < 0b10000000 then
    1
  else if n < 0b11100000 then
      2
    else if n < 0b11110000 then
        3
      else
        4
let is_ucschar c =
  let n = Uchar.to_int c in
  (n >= 0xA0 && n <= 0xD7FF) ||
    (n >= 0xF900 && n<= 0xFDCF) ||
    (n >= 0xFDF0 && n <= 0xFFEF) ||
    (n >= 0x10000 && n<= 0x1FFFD) ||
    (n >= 0x20000 && n<= 0x2FFFD) ||
    (n >= 0x30000 && n<= 0x3FFFD) ||
    (n >= 0x40000 && n<= 0x4FFFD) ||
    (n >= 0x50000 && n<= 0x5FFFD) ||
    (n >= 0x60000 && n<= 0x6FFFD) ||
    (n >= 0x70000 && n<= 0x7FFFD) ||
    (n >= 0x80000 && n<= 0x8FFFD) ||
    (n >= 0x90000 && n<= 0x9FFFD) ||
    (n >= 0xA0000 && n<= 0xAFFFD) ||
    (n >= 0xB0000 && n<= 0xBFFFD) ||
    (n >= 0xC0000 && n<= 0xCFFFD) ||
    (n >= 0xD0000 && n<= 0xDFFFD) ||
    (n >= 0xE1000 && n<= 0xEFFFD)
let is_iprivate c =
  let n = Uchar.to_int c in
  (n >= 0xE000 && n <= 0xF8FF) ||
    (n >= 0xF0000 && n <= 0xFFFFD) ||
    (n >= 0x100000 && n <= 0x10FFFD)

let pct_decode =
  let rec iter b len s i =
    if i >= len then
      ()
    else
      begin
        let i =
          match s.[i] with
            '%' when i+2 < len ->
              begin
                try
                  let n = int_of_string ("0x"^(String.sub s (i+1) 2)) in
                  let c = Char.chr n in
                  Buffer.add_char b c;
                  i+3
                with
                  _ -> Buffer.add_char b s.[i]; i+1
              end
          | _ ->
              let size = utf8_nb_bytes_of_char s.[i] in
              if size = 1 then
                Buffer.add_char b s.[i]
              else
                Buffer.add_substring b s i size ;
              i+size
        in
        iter b len s i
      end
  in
  fun s ->
    let len = String.length s in
    let b = Buffer.create len in
    iter b len s 0;
    Buffer.contents b
;;

let safe_chars =
  let f n =
    match Char.chr n with
      'A'..'Z' | 'a'..'z' | '0'..'9' | '-' | '.' -> true
    | _ -> false
  in
  Array.init 256 f
;;

let from_safe_chars ?(f=fun _ -> false) safe_chars c =
  let n = Uchar.to_int c in
  assert (n >= 0);
  if n > 255 then f c else safe_chars.(n)

let scheme_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code '+') <- true;
  a
;;
let scheme_safe_char = from_safe_chars scheme_safe_chars

let sub_delims =
  [| '!' ; '$' ; '&' ; '\'' ; '(' ; ')' ;
     '*' ; '+' ; ',' ; ';' ; '=' |]
;;

let user_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code ':') <- true;
  a.(Char.code '~') <- true;
  a.(Char.code '_') <- true;
  Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
  a
;;
let user_safe_char =
  from_safe_chars  ~f: is_ucschar user_safe_chars

let host_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code '~') <- true;
  a.(Char.code '_') <- true;
  a.(Char.code '[') <- true;
  a.(Char.code ']') <- true;
  a.(Char.code ':') <- true;
  Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
  a
;;
let host_safe_char =
  from_safe_chars ~f: is_ucschar host_safe_chars

let path_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code '~') <- true;
  a.(Char.code '_') <- true;
  a.(Char.code ':') <- true;
  a.(Char.code '@') <- true;
  Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
  a
;;
let path_safe_char =
  from_safe_chars ~f: is_ucschar path_safe_chars

let query_part_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code '~') <- true;
  a.(Char.code '_') <- true;
  a.(Char.code ':') <- true;
  a.(Char.code '@') <- true;
  a.(Char.code '?') <- true;
  a.(Char.code '/') <- true;
  Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
  a.(Char.code '=') <- false;
  a.(Char.code '&') <- false;
  a
;;
let query_part_safe_char =
  from_safe_chars ~f: is_ucschar query_part_safe_chars

let fragment_safe_chars =
  let a = Array.copy safe_chars in
  a.(Char.code '~') <- true;
  a.(Char.code '_') <- true;
  a.(Char.code ':') <- true;
  a.(Char.code '@') <- true;
  a.(Char.code '?') <- true;
  a.(Char.code '/') <- true;
  Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
  a
;;
let fragment_safe_char =
  from_safe_chars ~f: is_ucschar fragment_safe_chars

let pct_encode_utf8 dest codepoint =
  let b = Buffer.create 4 in
  Uutf.Buffer.add_utf_8 b codepoint;
  let s = Buffer.contents b in
  String.iter
    (fun c -> Printf.bprintf dest "%%%02X" (Char.code c))
    s

let pct_encode_b =
  let f is_safe_char b () _i = function
    `Malformed str -> Buffer.add_string b str
  | `Uchar codepoint ->
      if is_safe_char codepoint then
        Uutf.Buffer.add_utf_8 b codepoint
      else
        pct_encode_utf8 b codepoint
  in
  fun b is_safe_char s ->
    Uutf.String.fold_utf_8 (f is_safe_char b) () s
;;

let pct_encode is_safe_char s =
  let b = Buffer.create (String.length s) in
  pct_encode_b b is_safe_char s ;
  Buffer.contents b

let pct_encode_query =
  let safe_chars =
    let a = Array.copy safe_chars in
    a.(Char.code '~') <- true;
    a.(Char.code '_') <- true;
    a.(Char.code ':') <- true;
    a.(Char.code '@') <- true;
    a.(Char.code '?') <- true;
    a.(Char.code '/') <- true;
    Array.iter (fun c -> a.(Char.code c) <- true) sub_delims ;
    a
  in
  let is_safe_char =
    from_safe_chars ~f: is_ucschar safe_chars
  in
  fun str -> pct_encode is_safe_char str

let path_string =
  let string_of_path encode l =
    let l =
      if encode then
        List.map (pct_encode path_safe_char) l
      else
        l
    in
    String.concat "/" l
  in
  fun ?(pctencode=false) t ->
    match t.path with
      Absolute l -> "/"^(string_of_path pctencode l)
    | Relative l -> string_of_path pctencode l

let to_string =
  fun ?(pctencode=true) iri  ->
    let has_ihier = iri.host <> None in
    let b = Buffer.create 256 in
    if iri.scheme <> "" then
      (
       Buffer.add_string b iri.scheme ;
       Buffer.add_string b ":"
      );
    if has_ihier then Buffer.add_string b "//";
    (match iri.user with
       None -> ()
     | Some u ->
         Buffer.add_string b
           (if pctencode then pct_encode user_safe_char u else u);
         Buffer.add_char b '@'
    );
    (match iri.host with
       None -> ()
     | Some s ->
         Buffer.add_string b
           (if pctencode then pct_encode host_safe_char s else s)
    );
    (match iri.port with None -> () | Some n -> Buffer.add_string b (":"^(string_of_int n))) ;

    Buffer.add_string b (path_string ~pctencode iri);

    (match iri.query with
       None -> ()
     | Some s ->
         Buffer.add_char b '?';
         Buffer.add_string b s ;
    );
    (match iri.fragment with
       None -> ()
     | Some s ->
         Buffer.add_char b '#' ;
         Buffer.add_string b (if pctencode then pct_encode fragment_safe_char s else s)
    );
    Buffer.contents b
;;

let map_opt f = function None -> None | Some x -> Some (f x)

let utf8_split =
  let f split_char b_chars acc_words _i = function
    `Malformed str -> Buffer.add_string b_chars str; acc_words
  | `Uchar codepoint ->
      if split_char codepoint then
        begin
          let w = Buffer.contents b_chars in
          Buffer.reset b_chars ;
          match w with
            "" -> acc_words
          | _ -> w :: acc_words
        end
      else
        begin
          Uutf.Buffer.add_utf_8 b_chars codepoint ;
          acc_words
        end
  in
  fun split_char str ->
    let b = Buffer.create 128 in
    let words = Uutf.String.fold_utf_8 (f split_char b) [] str in
    match Buffer.contents b with
      "" -> List.rev words
    | w -> List.rev (w :: words)

let encode_query_string_part = pct_encode query_part_safe_char

let split_query_string =
  let dec = pct_decode in
  let cp_equal = Uchar.of_char '=' in
  let cp_amp = Uchar.of_char '&' in
  let is_equal = Uchar.equal cp_equal in
  let is_amp = Uchar.equal cp_amp in
  let add map str =
    match utf8_split is_equal str with
      [] | [_] -> KV.add (dec str) "" map
    | [ k ; v ] -> KV.add (dec k) (dec v) map
    | k :: vals -> KV.add (dec k) (dec (String.concat "=" vals)) map
  in
  fun str ->
    let l = utf8_split is_amp str in
    List.fold_left add KV.empty l

let split_query_opt = function
| None -> KV.empty
| Some q -> split_query_string q

let query_string_of_kv =
  let f b max_i i (k, v) =
     pct_encode_b b query_part_safe_char k ;
     Buffer.add_char b '=' ;
     pct_encode_b b query_part_safe_char v ;
     if i <> max_i then Buffer.add_char b '&'
  in
  fun ?q pairs ->
    let b = Buffer.create 128 in
    (match q with
       None | Some "" -> ()
     | Some str ->
         Buffer.add_string b str ;
         Buffer.add_char b '&'
    );
    let pairs = KV.bindings pairs in
    List.iteri (f b (List.length pairs - 1)) pairs ;
    Buffer.contents b

let query_string_of_kv_opt = map_opt query_string_of_kv

let iri ?(scheme="") ?user ?host ?port ?(path=Absolute[]) ?query_kv ?query ?fragment () =
  let (query, query_kv) =
    match query, query_kv with
    | None, None -> (None, None)
    | Some _, None -> (query, None)
    | None, Some kv -> (Some (query_string_of_kv kv), query_kv)
    | Some q, Some kv ->
        (* concat both and set kv pairs to none so that it will
           be recreated from the concatenation *)
        (Some (query_string_of_kv ~q kv), None)
  in
  { scheme ; user ; host ; port ; path ; query ; fragment ; query_kv }

let scheme t = t.scheme
let with_scheme t scheme = { t with scheme }

let user t = t.user
let with_user t user = { t with user }

let host t = t.host
let with_host t host = { t with host }

let port t = t.port
let with_port t port = { t with port }

let path t = t.path
let with_path t path = { t with path }

let append_path t strings =
  let l = match t.path with
    Absolute l | Relative l -> l
  in
  let l =
    match List.rev l with
      "" :: q -> (List.rev q) @ strings
    | _ -> l @ strings
  in
  let path = match t.path with
    | Absolute _-> Absolute l
    | Relative _ -> Relative l
  in
  with_path t path

let query t = t.query
let with_query t query = { t with query }

let query_kv t =
  match t.query_kv with
    Some map -> map
  | None ->
      let map = split_query_opt t.query in
      t.query_kv <- Some map;
      map

let with_query_kv t map =
  if KV.is_empty map then
    { t with query = None ; query_kv = None }
  else
    { t with
      query = Some (query_string_of_kv map) ;
      query_kv = Some map ;
    }

let query_get t key =
  match KV.find key (query_kv t) with
  | exception Not_found -> ""
  | str -> str

let query_opt t key =
  match KV.find key (query_kv t) with
  | exception Not_found -> None
  | str -> Some str

let query_set t k v =
  let map = query_kv t in
  let map = KV.add k v map in
  with_query_kv t map

let fragment t = t.fragment
let with_fragment t fragment = { t with fragment }

let compare i1 i2 =
  ignore(query_kv i1);
  ignore(query_kv i2);
  Pervasives.compare i1 i2

let equal i1 i2 = compare i1 i2 = 0

let normalize_path =
  let rec iter acc abs = function
    [] -> List.rev acc
  | "." :: q -> iter acc abs (""::q)
  | ".." :: q ->
      begin
        match acc with
          [] -> iter acc abs q
        | _ :: acc -> iter acc abs (""::q)
      end
  | [""] -> List.rev (""::acc)
  | "" :: q -> iter acc abs q
  | h :: q -> iter (h :: acc) abs q
  in
  iter [] true
;;

let path_remove_dot_segments = function
  Absolute l -> Absolute (normalize_path l)
| Relative l -> Relative (normalize_path l)

let remove_dot_segments t =
  let path = path_remove_dot_segments t.path in
  { t with path }

let normalize_host s =
  let len = String.length s in
  if len > 0 then
    match String.get s 0 with
      '[' -> String.uppercase_ascii s (* uppercase hexa *)
    | _ -> String.lowercase_ascii s (* lowercase regname *)
  else
    s

let normalize_port t =
  match String.lowercase_ascii t.scheme, t.port with
    "http", Some 80 -> { t with port = None }
  | "https", Some 443 -> { t with port = None }
  | "ftp", Some 21 -> { t with port = None }
  | "sftp", Some 115 -> { t with port = None }
  | "ssh", Some 22 -> { t with port = None }
  | "smtp", Some 25 -> { t with port = None }
  | _ -> t

let normalize_case t =
  { t with
    scheme = String.lowercase_ascii t.scheme ;
    host = map_opt normalize_host t.host ;
  }

let normalize_nfkc t =
  let f = Uunf_string.normalize_utf_8 `NFKC in
  let path =
    match t.path with
      Absolute l -> Absolute (List.map f l)
    | Relative l -> Relative (List.map f l)
  in
  { t with
    host = map_opt f t.host ;
    path ;
    user = map_opt f t.user ;
    query = map_opt f t.query ; (* beware: the query is not %-decoded *)
    fragment = map_opt f t.fragment ;
    query_kv = None ;
  }

let normalize ?(nfkc=true) t =
  let t = remove_dot_segments t in
  let t = normalize_case t in
  let t = normalize_port t in
  if nfkc then normalize_nfkc t else t

let to_uri =
  let user_safe_char = from_safe_chars user_safe_chars in
  let host_safe_char = from_safe_chars host_safe_chars in
  let fragment_safe_char = from_safe_chars fragment_safe_chars in
  let query_part_safe_char = from_safe_chars query_part_safe_chars in
  let path_safe_char = from_safe_chars path_safe_chars in
  let path_string =
    let string_of_path l =
      let l = List.map (pct_encode path_safe_char) l in
      String.concat "/" l
    in
    fun t ->
      match t.path with
        Absolute l -> "/"^(string_of_path l)
      | Relative l -> string_of_path l
  in
  let print_query_string =
    let f b max_i i (k, v) =
      pct_encode_b b query_part_safe_char k ;
      Buffer.add_char b '=' ;
      pct_encode_b b query_part_safe_char v ;
      if i <> max_i then Buffer.add_char b '&'
    in
    fun b t ->
      match KV.bindings (query_kv t) with
      | [] -> ()
      | pairs ->
          Buffer.add_char b '?';
          List.iteri (f b (List.length pairs - 1)) pairs
  in
  fun iri  ->
    let has_ihier = iri.host <> None in
    let b = Buffer.create 256 in
    if iri.scheme <> "" then
      (
       Buffer.add_string b iri.scheme ;
       Buffer.add_string b ":"
      );
    if has_ihier then Buffer.add_string b "//";
    (match iri.user with
       None -> ()
     | Some u ->
         pct_encode_b b user_safe_char u ;
         Buffer.add_char b '@'
    );
    (match iri.host with
       None -> ()
     | Some s -> pct_encode_b b host_safe_char s
    );
    (match iri.port with
     | None -> ()
     | Some n -> Buffer.add_string b (":"^(string_of_int n))
    ) ;

    Buffer.add_string b (path_string iri);

    print_query_string b iri ;

    (match iri.fragment with
       None -> ()
     | Some s ->
         Buffer.add_char b '#' ;
         pct_encode_b b fragment_safe_char s
    );
    Buffer.contents b
;;


